istioctl/cmd/revision.go

```diff
@@ -0,0 +1,558 @@
+// limitations under the License.
+
+package cmd
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"sort"
+	"strings"
+	"text/tabwriter"
+	"time"
+
+	"github.com/spf13/cobra"
+	v1 "k8s.io/api/core/v1"
+	meta_v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	apimachinery_schema "k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/duration"
+	"k8s.io/apimachinery/pkg/util/validation"
+	"k8s.io/cli-runtime/pkg/genericclioptions"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/utils/pointer"
+
+	"istio.io/api/label"
+	"istio.io/api/operator/v1alpha1"
+	"istio.io/istio/operator/cmd/mesh"
+	operator_istio "istio.io/istio/operator/pkg/apis/istio"
+	iopv1alpha1 "istio.io/istio/operator/pkg/apis/istio/v1alpha1"
+	"istio.io/istio/operator/pkg/manifest"
+	"istio.io/istio/operator/pkg/util"
+	"istio.io/istio/operator/pkg/util/clog"
+	"istio.io/istio/pkg/config"
+)
+
+type revisionArgs struct {
+	manifestsPath string
+	name          string
+	verbose       bool
+}
+
+var (
+	istioOperatorGVR = apimachinery_schema.GroupVersionResource{
+		Group:    iopv1alpha1.SchemeGroupVersion.Group,
+		Version:  iopv1alpha1.SchemeGroupVersion.Version,
+		Resource: "istiooperators",
+	}
+
+	kubeConfigFlags = &genericclioptions.ConfigFlags{
+		Context:    pointer.StringPtr(""),
+		Namespace:  pointer.StringPtr(""),
+		KubeConfig: pointer.StringPtr(""),
+	}
+
+	revArgs = revisionArgs{}
+)
+
+func revisionCommand() *cobra.Command {
+	revisionCmd := &cobra.Command{
+		Use:     "revision",
+		Short:   "Revision centric view of Istio deployment",
+		Aliases: []string{"rev"},
+	}
+	revisionCmd.PersistentFlags().StringVarP(&revArgs.manifestsPath, "manifests", "d", "", mesh.ManifestsFlagHelpStr)
+	revisionCmd.PersistentFlags().BoolVarP(&revArgs.verbose, "verbose", "v", false, "print customizations")
+
+	revisionCmd.AddCommand(revisionListCommand())
+	revisionCmd.AddCommand(revisionDescribeCommand())
+	return revisionCmd
+}
+
+func revisionDescribeCommand() *cobra.Command {
+	describeCmd := &cobra.Command{
+		Use:     "describe",
+		Example: `    istioctl experimental revision describe canary`,
+		Short:   "Show details of a revision - customizations, number of pods pointing to it, istiod, gateways etc",
+		Args: func(cmd *cobra.Command, args []string) error {
+			if len(args) == 0 {
+				return fmt.Errorf("revision is not specified")
+			}
+			if len(args) > 1 {
+				return fmt.Errorf("exactly 1 revision should be specified")
+			}
+			return nil
+		},
+		RunE: func(cmd *cobra.Command, args []string) error {
+			revArgs.name = args[0]
+			if errs := validation.IsDNS1123Label(revArgs.name); len(errs) > 0 {
+				return fmt.Errorf(strings.Join(errs, "\n"))
+			}
+			logger := clog.NewConsoleLogger(cmd.OutOrStdout(), cmd.ErrOrStderr(), scope)
+			return revisionDescription(cmd.OutOrStdout(), &revArgs, kubeConfigFlags, logger)
+		},
+	}
+	describeCmd.Flags().Bool("all", true, "show all related to a revision")
```

Comment:
"all" is not commonly used in istioctl - can this be accommodated with verbose?

Reply:
Yeah. Fixed that.